In [1]:
#importing necessary labraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
#reading dataframe
df = pd.read_csv(r"C:\Users\Roshan Salunke\Downloads\Data Science Course\ml project\9\transaction_dataset.csv")

In [3]:
#checking head of the data
df.head()

Unnamed: 0  Index                                     Address  FLAG  \
0           0      1  0x00009277775ac7d0d59eaad8fee3d10ac6c805e8     0   
1           1      2  0x0002b44ddb1476db43c868bd494422ee4c136fed     0   
2           2      3  0x0002bda54cb772d040f779e88eb453cac0daa244     0   
3           3      4  0x00038e6ba2fd5c09aedb96697c8d7b8fa6632e5e     0   
4           4      5  0x00062d1dd1afb6fb02540ddad9cdebfe568e0d89     0   

   Avg min between sent tnx  Avg min between received tnx  \
0                    844.26                       1093.71   
1                  12709.07                       2958.44   
2                 246194.54                       2434.02   
3                  10219.60                      15785.09   
4                     36.61                      10707.77   

   Time Diff between first and last (Mins)  Sent tnx  Received Tnx  \
0                                704785.63       721            89   
1                               1218216.73        94             8   
2                                516729.30         2            10   
3                                397555.90        25             9   
4                                382472.42      4598            20   

   Number of Created Contracts  ...   ERC20 min val sent   ERC20 max val sent  \
0                            0  ...             0.000000         1.683100e+07   
1                            0  ...             2.260809         2.260809e+00   
2                            0  ...             0.000000         0.000000e+00   
3                            0  ...           100.000000         9.029231e+03   
4                            1  ...             0.000000         4.500000e+04   

    ERC20 avg val sent   ERC20 min val sent contract  \
0        271779.920000                           0.0   
1             2.260809                           0.0   
2             0.000000                           0.0   
3          3804.076893                           0.0   
4         13726.659220                           0.0   

    ERC20 max val sent contract   ERC20 avg val sent contract  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

    ERC20 uniq sent token name   ERC20 uniq rec token name  \
0                         39.0                        57.0   
1                          1.0                         7.0   
2                          0.0                         8.0   
3                          1.0                        11.0   
4                          6.0                        27.0   

    ERC20 most sent token type   ERC20_most_rec_token_type  
0                    Cofoundit                   Numeraire  
1               Livepeer Token              Livepeer Token  
2                         None                       XENON  
3                       Raiden                       XENON  
4                StatusNetwork                         EOS  

[5 rows x 51 columns]

In [4]:
#checking if of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9841 entries, 0 to 9840
Data columns (total 51 columns):
 #   Column                                                Non-Null Count  Dtype  
---  ------                                                --------------  -----  
 0   Unnamed: 0                                            9841 non-null   int64  
 1   Index                                                 9841 non-null   int64  
 2   Address                                               9841 non-null   object 
 3   FLAG                                                  9841 non-null   int64  
 4   Avg min between sent tnx                              9841 non-null   float64
 5   Avg min between received tnx                          9841 non-null   float64
 6   Time Diff between first and last (Mins)               9841 non-null   float64
 7   Sent tnx                                              9841 non-null   int64  
 8   Received Tnx                                          9841

Total 51 columns and 9012 rows with 3 object columns and 48 numerical.

In [5]:
#checking statistical info of data
df.describe()

Unnamed: 0        Index         FLAG  Avg min between sent tnx  \
count  9841.000000  9841.000000  9841.000000               9841.000000   
mean   4920.000000  1815.049893     0.221421               5086.878721   
std    2840.996333  1222.621830     0.415224              21486.549974   
min       0.000000     1.000000     0.000000                  0.000000   
25%    2460.000000   821.000000     0.000000                  0.000000   
50%    4920.000000  1641.000000     0.000000                 17.340000   
75%    7380.000000  2601.000000     0.000000                565.470000   
max    9840.000000  4729.000000     1.000000             430287.670000   

       Avg min between received tnx  Time Diff between first and last (Mins)  \
count                   9841.000000                             9.841000e+03   
mean                    8004.851184                             2.183333e+05   
std                    23081.714801                             3.229379e+05   
min                        0.000000                             0.000000e+00   
25%                        0.000000                             3.169300e+02   
50%                      509.770000                             4.663703e+04   
75%                     5480.390000                             3.040710e+05   
max                   482175.490000                             1.954861e+06   

           Sent tnx  Received Tnx  Number of Created Contracts  \
count   9841.000000   9841.000000                  9841.000000   
mean     115.931714    163.700945                     3.729702   
std      757.226361    940.836550                   141.445583   
min        0.000000      0.000000                     0.000000   
25%        1.000000      1.000000                     0.000000   
50%        3.000000      4.000000                     0.000000   
75%       11.000000     27.000000                     0.000000   
max    10000.000000  10000.000000                  9995.000000   

       Unique Received From Addresses  ...   ERC20 max val rec  \
count                     9841.000000  ...        9.012000e+03   
mean                        30.360939  ...        1.252524e+08   
std                        298.621112  ...        1.053741e+10   
min                          0.000000  ...        0.000000e+00   
25%                          1.000000  ...        0.000000e+00   
50%                          2.000000  ...        0.000000e+00   
75%                          5.000000  ...        9.900000e+01   
max                       9999.000000  ...        1.000000e+12   

        ERC20 avg val rec   ERC20 min val sent   ERC20 max val sent  \
count        9.012000e+03         9.012000e+03         9.012000e+03   
mean         4.346203e+06         1.174126e+04         1.303594e+07   
std          2.141192e+08         1.053567e+06         1.179905e+09   
min          0.000000e+00         0.000000e+00         0.000000e+00   
25%          0.000000e+00         0.000000e+00         0.000000e+00   
50%          0.000000e+00         0.000000e+00         0.000000e+00   
75%          2.946467e+01         0.000000e+00         0.000000e+00   
max          1.724181e+10         1.000000e+08         1.120000e+11   

        ERC20 avg val sent   ERC20 min val sent contract  \
count         9.012000e+03                        9012.0   
mean          6.318389e+06                           0.0   
std           5.914764e+08                           0.0   
min           0.000000e+00                           0.0   
25%           0.000000e+00                           0.0   
50%           0.000000e+00                           0.0   
75%           0.000000e+00                           0.0   
max           5.614756e+10                           0.0   

        ERC20 max val sent contract   ERC20 avg val sent contract  \
count                        9012.0                        9012.0   
mean                            0.0                           0.0   
std                             0.0       

In [6]:
#checking the unique values in the dataframe
for i in df.columns:
    uni = df[i].nunique()
    print(i,' :',uni)

Unnamed: 0  : 9841
Index  : 4729
Address  : 9816
FLAG  : 2
Avg min between sent tnx  : 5013
Avg min between received tnx  : 6223
Time Diff between first and last (Mins)  : 7810
Sent tnx  : 641
Received Tnx  : 727
Number of Created Contracts  : 20
Unique Received From Addresses  : 256
Unique Sent To Addresses  : 258
min value received  : 4589
max value received   : 6302
avg val received  : 6767
min val sent  : 4719
max val sent  : 6647
avg val sent  : 5854
min value sent to contract  : 3
max val sent to contract  : 4
avg value sent to contract  : 4
total transactions (including tnx to create contract  : 897
total Ether sent  : 5868
total ether received  : 6728
total ether sent contracts  : 4
total ether balance  : 5717
 Total ERC20 tnxs  : 300
 ERC20 total Ether received  : 3460
 ERC20 total ether sent  : 1415
 ERC20 total Ether sent contract  : 29
 ERC20 uniq sent addr  : 107
 ERC20 uniq rec addr  : 147
 ERC20 uniq sent addr.1  : 4
 ERC20 uniq rec contract addr  : 123
 ERC20 avg time b

In [7]:
#dropping first two columns, as they are index columns
df.drop(['Unnamed: 0','Index'],axis=1,inplace=True)

In [8]:
df.head()

Address  FLAG  Avg min between sent tnx  \
0  0x00009277775ac7d0d59eaad8fee3d10ac6c805e8     0                    844.26   
1  0x0002b44ddb1476db43c868bd494422ee4c136fed     0                  12709.07   
2  0x0002bda54cb772d040f779e88eb453cac0daa244     0                 246194.54   
3  0x00038e6ba2fd5c09aedb96697c8d7b8fa6632e5e     0                  10219.60   
4  0x00062d1dd1afb6fb02540ddad9cdebfe568e0d89     0                     36.61   

   Avg min between received tnx  Time Diff between first and last (Mins)  \
0                       1093.71                                704785.63   
1                       2958.44                               1218216.73   
2                       2434.02                                516729.30   
3                      15785.09                                397555.90   
4                      10707.77                                382472.42   

   Sent tnx  Received Tnx  Number of Created Contracts  \
0       721            89                            0   
1        94             8                            0   
2         2            10                            0   
3        25             9                            0   
4      4598            20                            1   

   Unique Received From Addresses  Unique Sent To Addresses  ...  \
0                              40                       118  ...   
1                               5                        14  ...   
2                              10                         2  ...   
3                               7                        13  ...   
4                               7                        19  ...   

    ERC20 min val sent   ERC20 max val sent   ERC20 avg val sent  \
0             0.000000         1.683100e+07        271779.920000   
1             2.260809         2.260809e+00             2.260809   
2             0.000000         0.000000e+00             0.000000   
3           100.000000         9.029231e+03          3804.076893   
4             0.000000         4.500000e+04         13726.659220   

    ERC20 min val sent contract   ERC20 max val sent contract  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

    ERC20 avg val sent contract   ERC20 uniq sent token name  \
0                           0.0                         39.0   
1                           0.0                          1.0   
2                           0.0                          0.0   
3                           0.0                          1.0   
4                           0.0                          6.0   

    ERC20 uniq rec token name   ERC20 most sent token type  \
0                        57.0                    Cofoundit   
1                         7.0               Livepeer Token   
2                         8.0                         None   
3                        11.0                       Raiden   
4                        27.0                StatusNetwork   

    ERC20_most_rec_token_type  
0                   Numeraire  
1              Livepeer Token  
2                       XENON  
3                       XENON  
4                         EOS  

[5 rows x 49 columns]

In [9]:
#checking the null values
df.isnull().sum()

Address                                                   0
FLAG                                                      0
Avg min between sent tnx                                  0
Avg min between received tnx                              0
Time Diff between first and last (Mins)                   0
Sent tnx                                                  0
Received Tnx                                              0
Number of Created Contracts                               0
Unique Received From Addresses                            0
Unique Sent To Addresses                                  0
min value received                                        0
max value received                                        0
avg val received                                          0
min val sent                                              0
max val sent                                              0
avg val sent                                              0
min value sent to contract              

1. There few columns with null values.
2. We either drop them or replace them

We are replace null values with median, because we are doing classification, so no need to check the distribution of the values. We can directly imput missing values with median.

In [10]:
# Replacing missing numerical values with median
for i in df.columns:
    if df[i].dtype == 'int64' or df[i].dtype == 'float64':
        df[i].fillna(df[i].median(),inplace=True)

In [11]:
#checking the null values again. only categorical columns are having null values
df.isnull().sum()

Address                                                   0
FLAG                                                      0
Avg min between sent tnx                                  0
Avg min between received tnx                              0
Time Diff between first and last (Mins)                   0
Sent tnx                                                  0
Received Tnx                                              0
Number of Created Contracts                               0
Unique Received From Addresses                            0
Unique Sent To Addresses                                  0
min value received                                        0
max value received                                        0
avg val received                                          0
min val sent                                              0
max val sent                                              0
avg val sent                                              0
min value sent to contract              

In [12]:
#checking unique values for categorical columns
for i in df.columns:
    if df[i].dtype =='object':
        uni = df[i].nunique()
        print(i,':',uni)

Address : 9816
 ERC20 most sent token type : 305
 ERC20_most_rec_token_type : 467


In [13]:
#replacing missing values with mode
for i in df.columns:
    if df[i].dtype=='object':
        df[i].fillna(df[i].mode()[0],inplace=True)

In [14]:
#checking null values
for i in df.columns:
    null = df[i].isnull().sum()
    print(i,':',null)

Address : 0
FLAG : 0
Avg min between sent tnx : 0
Avg min between received tnx : 0
Time Diff between first and last (Mins) : 0
Sent tnx : 0
Received Tnx : 0
Number of Created Contracts : 0
Unique Received From Addresses : 0
Unique Sent To Addresses : 0
min value received : 0
max value received  : 0
avg val received : 0
min val sent : 0
max val sent : 0
avg val sent : 0
min value sent to contract : 0
max val sent to contract : 0
avg value sent to contract : 0
total transactions (including tnx to create contract : 0
total Ether sent : 0
total ether received : 0
total ether sent contracts : 0
total ether balance : 0
 Total ERC20 tnxs : 0
 ERC20 total Ether received : 0
 ERC20 total ether sent : 0
 ERC20 total Ether sent contract : 0
 ERC20 uniq sent addr : 0
 ERC20 uniq rec addr : 0
 ERC20 uniq sent addr.1 : 0
 ERC20 uniq rec contract addr : 0
 ERC20 avg time between sent tnx : 0
 ERC20 avg time between rec tnx : 0
 ERC20 avg time between rec 2 tnx : 0
 ERC20 avg time between contract tnx

In [15]:
#using label encoding for categorical columns
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
for i in df.columns:
    if df[i].dtype=='object':
        df[i] = le.fit_transform(df[i])

In [16]:
df.head()

Address  FLAG  Avg min between sent tnx  Avg min between received tnx  \
0        0     0                    844.26                       1093.71   
1        1     0                  12709.07                       2958.44   
2        2     0                 246194.54                       2434.02   
3        3     0                  10219.60                      15785.09   
4        4     0                     36.61                      10707.77   

   Time Diff between first and last (Mins)  Sent tnx  Received Tnx  \
0                                704785.63       721            89   
1                               1218216.73        94             8   
2                                516729.30         2            10   
3                                397555.90        25             9   
4                                382472.42      4598            20   

   Number of Created Contracts  Unique Received From Addresses  \
0                            0                              40   
1                            0                               5   
2                            0                              10   
3                            0                               7   
4                            1                               7   

   Unique Sent To Addresses  ...   ERC20 min val sent   ERC20 max val sent  \
0                       118  ...             0.000000         1.683100e+07   
1                        14  ...             2.260809         2.260809e+00   
2                         2  ...             0.000000         0.000000e+00   
3                        13  ...           100.000000         9.029231e+03   
4                        19  ...             0.000000         4.500000e+04   

    ERC20 avg val sent   ERC20 min val sent contract  \
0        271779.920000                           0.0   
1             2.260809                           0.0   
2             0.000000                           0.0   
3          3804.076893                           0.0   
4         13726.659220                           0.0   

    ERC20 max val sent contract   ERC20 avg val sent contract  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

    ERC20 uniq sent token name   ERC20 uniq rec token name  \
0                         39.0                        57.0   
1                          1.0                         7.0   
2                          0.0                         8.0   
3                          1.0                        11.0   
4                          6.0                        27.0   

    ERC20 most sent token type   ERC20_most_rec_token_type  
0                           63                         300  
1                          154                         257  
2                          185                         436  
3                          217                         436  
4                          238                         154  

[5 rows x 49 columns]

In [17]:
#saperating features
x = df.drop('FLAG',axis=1)
y = df['FLAG']

In [18]:
#scaling the dataset
from sklearn.preprocessing import StandardScaler
xcols = x.columns
sc = StandardScaler()
x = sc.fit_transform(x)
x = pd.DataFrame(x,columns=xcols)
x.head()

Address  Avg min between sent tnx  Avg min between received tnx  \
0 -1.732447                 -0.197465                     -0.299436   
1 -1.732094                  0.354760                     -0.218644   
2 -1.731740                 11.221900                     -0.241365   
3 -1.731387                  0.238893                      0.337091   
4 -1.731034                 -0.235055                      0.117108   

   Time Diff between first and last (Mins)  Sent tnx  Received Tnx  \
0                                 1.506411  0.799099     -0.079402   
1                                 3.096367 -0.028965     -0.165500   
2                                 0.924051 -0.150467     -0.163375   
3                                 0.555004 -0.120091     -0.164437   
4                                 0.508294  5.919361     -0.152745   

   Number of Created Contracts  Unique Received From Addresses  \
0                     -0.02637                        0.032280   
1                     -0.02637                       -0.084931   
2                     -0.02637                       -0.068187   
3                     -0.02637                       -0.078233   
4                     -0.01930                       -0.078233   

   Unique Sent To Addresses  min value received  ...   ERC20 min val sent  \
0                  0.349346           -0.134530  ...            -0.010665   
1                 -0.044882           -0.134530  ...            -0.010663   
2                 -0.090370           -0.134183  ...            -0.010665   
3                 -0.048673           -0.134530  ...            -0.010566   
4                 -0.025929           -0.134530  ...            -0.010665   

    ERC20 max val sent   ERC20 avg val sent   ERC20 min val sent contract  \
0             0.004334            -0.009743                           0.0   
1            -0.010573            -0.010223                           0.0   
2            -0.010573            -0.010223                           0.0   
3            -0.010565            -0.010216                           0.0   
4            -0.010533            -0.010199                           0.0   

    ERC20 max val sent contract   ERC20 avg val sent contract  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

    ERC20 uniq sent token name   ERC20 uniq rec token name  \
0                     5.844163                    3.281981   
1                    -0.041551                    0.156028   
2                    -0.196438                    0.218547   
3                    -0.041551                    0.406104   
4                     0.732885                    1.406409   

    ERC20 most sent token type   ERC20_most_rec_token_type  
0                    -0.004890                    1.376941  
1                     1.012303                    1.071381  
2                     1.358819                    2.343361  
3                     1.716514                    2.343361  
4                     1.951251                    0.339460  

[5 rows x 48 columns]

In [19]:
#spliting the data for train and test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test  = train_test_split(x,y,test_size=0.3,random_state=1)

Model Building

In [20]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

In [21]:
pred = lr.predict(x_test)

In [22]:
from sklearn.metrics import accuracy_score, classification_report
accuracy_score(y_test,pred)

0.8807991872671859

In [23]:
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.87      0.99      0.93      2306
           1       0.93      0.50      0.65       647

    accuracy                           0.88      2953
   macro avg       0.90      0.74      0.79      2953
weighted avg       0.89      0.88      0.87      2953



In [24]:
from sklearn.svm import LinearSVC
svm = LinearSVC()
svm.fit(x_train,y_train)

LinearSVC()

In [25]:
pred = svm.predict(x_test)

In [26]:
print(classification_report(y_test,pred))

              precision    recall  f1-score   support

           0       0.85      0.99      0.91      2306
           1       0.90      0.36      0.51       647

    accuracy                           0.85      2953
   macro avg       0.87      0.67      0.71      2953
weighted avg       0.86      0.85      0.82      2953



In [27]:
from sklearn.ensemble import GradientBoostingClassifier
gb = GradientBoostingClassifier()
gb.fit(x_train,y_train)

GradientBoostingClassifier()

In [28]:
pred = gb.predict(x_test)

In [29]:
print(classification_report(y_test,pred))

              precision    recall  f1-score   support

           0       0.99      1.00      0.99      2306
           1       0.99      0.96      0.98       647

    accuracy                           0.99      2953
   macro avg       0.99      0.98      0.99      2953
weighted avg       0.99      0.99      0.99      2953



In [30]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dtmodel = dt.fit(x_train, y_train)

In [31]:
pred = dtmodel.predict(x_test)

In [32]:
print(classification_report(y_test,pred))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      2306
           1       0.97      0.97      0.97       647

    accuracy                           0.99      2953
   macro avg       0.98      0.98      0.98      2953
weighted avg       0.99      0.99      0.99      2953



In [33]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=42)
rf.fit(x_train,y_train)

RandomForestClassifier()

In [34]:
pred = rf.predict(x_test)

In [35]:
print(classification_report(y_test,pred))

              precision    recall  f1-score   support

           0       0.99      1.00      0.99      2306
           1       1.00      0.96      0.98       647

    accuracy                           0.99      2953
   macro avg       0.99      0.98      0.99      2953
weighted avg       0.99      0.99      0.99      2953



In [36]:
predtrain = rf.predict(x_train)
accuracy_score(y_train, predtrain)

1.0

As you can see, the training accuracy of the model is 100% and testing accuracy is 99%.

Hyperparameter Tuning

Our most of the models are performing well, we can do hyperparameter tuning for one model.

In [50]:
from sklearn.model_selection import RandomizedSearchCV
rf = RandomForestClassifier(random_state=42)
parameters = {
    'n_estimators': [50, 100, 200, 400]
}

sear = RandomizedSearchCV(estimator=rf,
                         param_distributions=parameters,
                         cv=5,
                         return_train_score=True)
sear.fit(x, y)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
                   param_distributions={'n_estimators': [50, 100, 200, 400]},
                   return_train_score=True)

In [51]:
sear.best_params_

{'n_estimators': 50}

In [52]:
parameters = {
    'criterion': ["gini", "entropy", "log_loss"]
}
    
sear = RandomizedSearchCV(estimator=rf,
                         param_distributions=parameters,
                         cv=5,
                         return_train_score=True)
sear.fit(x, y)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
                   param_distributions={'criterion': ['gini', 'entropy',
                                                      'log_loss']},
                   return_train_score=True)

In [53]:
sear.best_params_

{'criterion': 'entropy'}

In [54]:
parameters = {
    'max_depth': [1, 2, 3, 4]}
sear = RandomizedSearchCV(estimator=rf,
                         param_distributions=parameters,
                         cv=5,
                         return_train_score=True)
sear.fit(x, y)
sear.best_params_

{'max_depth': 4}

In [55]:
parameters = {
    'min_samples_split': [1, 2, 3, 4]}
sear = RandomizedSearchCV(estimator=rf,
                         param_distributions=parameters,
                         cv=5,
                         return_train_score=True)
sear.fit(x, y)
sear.best_params_

{'min_samples_split': 3}

In [56]:
parameters = {
    'min_samples_leaf' : [1, 2, 3, 4, 5]}
sear = RandomizedSearchCV(estimator=rf,
                         param_distributions=parameters,
                         cv=5,
                         return_train_score=True)
sear.fit(x, y)
sear.best_params_

{'min_samples_leaf': 5}

In [57]:
parameters = {
    'ccp_alpha': [0.001, 0.01, 0.1, 1]}
sear = RandomizedSearchCV(estimator=rf,
                         param_distributions=parameters,
                         cv=5,
                         return_train_score=True)
sear.fit(x, y)
sear.best_params_

{'ccp_alpha': 0.001}

In [58]:
rf = RandomForestClassifier(n_estimators=50,
                           max_depth=4,
                           min_samples_split=3,
                           min_samples_leaf=5,
                           ccp_alpha=0.001,
                           criterion='entropy')
rfmodel = rf.fit(x_train, y_train)

In [59]:
predtest  = rfmodel.predict(x_test)

In [60]:
predtrain = rfmodel.predict(x_train)

In [61]:
print(f'Accuracy score of traning is: {accuracy_score(predtrain, y_train)}')
print(f'Accuracy score of testing is: {accuracy_score(predtest, y_test)}')

Accuracy score of traning is: 0.9834494773519163
Accuracy score of testing is: 0.9817135116830342


In classification, we accuracy score is not that important. 
We should check confusion metrix and classification report

In [62]:
print(classification_report(y_test, predtest))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99      2306
           1       0.99      0.92      0.96       647

    accuracy                           0.98      2953
   macro avg       0.99      0.96      0.97      2953
weighted avg       0.98      0.98      0.98      2953



1. The precision is 0.98
2. The recall is 1.00
3. The f1-score is 0.99
4. Overall our result is very good. f1-score is the weighted average of precision and recall. The higher the fi1-score, the better is the model.